controller/src/main/java/io/pravega/controller/server/SegmentStoreConnectionManager.java

```diff
@@ -0,0 +1,653 @@
+                    waitQueue.add(request);
+                    connectionFuture = request.getFuture();
+                    tryCreateNewConnection = true;
+                }
+            }
+            if (tryCreateNewConnection) {
+                tryCreateNewConnection();
+            }
+            return connectionFuture;
+        }
+
+        /**
+         * Users use this method to return the connection back to the Connection Manager for reuse.
+         * This method first checks the state of the connection and if the connection is disconnected, it is closed.
+         * For valid connections, it checks if there is a waiting request for a connection and fulfils that first.
+         * If there are no waiting requests, it checks if the available queue has the capacity to keep this connection.
+         * If not, it destroys this connection.
+         *
+         * @param connectionObject connection to return to the pool.
+         */
+        void returnConnection(ConnectionObject connectionObject) {
+            connectionObject.reusableReplyProcessor.uninitialize();
+            if (connectionObject.state.get().equals(ConnectionObject.ConnectionState.DISCONNECTED)) {
+                handleDisconnected(connectionObject);
+            } else {
+                boolean toClose = false;
+                synchronized (lock) {
+                    WaitingRequest waiting = waitQueue.poll();
+                    if (waiting != null) {
+                        connectionObject.reusableReplyProcessor.initialize(waiting.getReplyProcessor());
+                        waiting.getFuture().complete(connectionObject);
+                    } else {
+                        if (!isRunning) {
+                            // The connection will be closed if returned anytime after the shutdown has been initiated.
+                            log.debug("ConnectionManager is shutdown");
+                            connectionCount--;
+                            toClose = true;
+                        } else {
+                            // as connections are returned to us, we put them in queue to be reused
+                            if (availableConnections.size() < maxAvailableConnections) {
+                                // if returned connection increases our available connection count, do not include it
+                                log.debug("Returned connection object is included in the available list");
+                                availableConnections.offer(connectionObject);
+                            } else {
+                                log.debug("Returned connection object is discarded as available list is full");
+                                connectionCount--;
+                                toClose = true;
+                            }
+                        }
+                    }
+                }
+
+                if (toClose) {
+                    if (connectionListener != null) {
```

Comment:
Why would `connectionListener` be null? Don't you pass it in through the constructor?

Reply:
its passed through a constructor thats only created for testing. 
production code does pass null